src/lib.rs

```diff
@@ -39,8 +39,48 @@
 #![crate_type = "rlib"]
 #![feature(core_intrinsics)]
 
+use std::{
+    alloc::{Layout, GlobalAlloc},
+    ops::Deref,
+    sync::atomic::{AtomicBool, Ordering},
+};
 
 static SIZE_ALLOC_INFO: usize = (1024 * 1024) * 2; // 2KB
+/// A spinlock for the global ALLOC_INFO list.
+///
+/// System allocators are generally thread-safe, but gcmalloc reads and writes
+/// metadata about each allocation to shared memory upon returning from the
+/// system allocator call. This requires additional synchronisation mechanisms.
+/// It is not possible to use `sys::Sync::Mutex` for this purpose as the
+/// implementation includes a heap allocation containing a `pthread_mutex`. This
+/// would cause infinite recursion when initialising the lock as the allocator
+/// would call itself.
+///
+/// Instead, we use a simple global spinlock built on an atomic bool to guard
+/// access to the ALLOC_INFO list.
+static ALLOC_LOCK: AllocLock = AllocLock(AtomicBool::new(false));
+
+struct AllocLock(AtomicBool);
+
+impl AllocLock {
+    fn lock(&self) {
+        while !ALLOC_LOCK.compare_and_swap(false, true, Ordering::AcqRel) {
+            // Spin
+        }
+    }
+
+    fn unlock(&self) {
+            ALLOC_LOCK.store(false, Ordering::Release);
+    }
+}
+
+impl Deref for AllocLock {
```

Comment:
Why would anyone deref `AllocLock`?

Reply:
This just lets you call `.lock()` `.unlock()` on `AllocLock` and have them update the atomic directly. It's just a bit more ergonomic than top level fn declarations.

Reply:
... But I've just noticed that I've forgotten to make `lock()` and `unlock() reference `self`. Which is why this might have been non-obvious!

Reply:
I'm not sure I understand. Couldn't `AllocLock.lock` have code along the lines of `self.0.compare_and_swap` and then you can remove this `Deref` implementation?

Reply:
Yes. This was more useful when I implemented an RAII guard (similar to `sys::Sync::MutexGuard`) so that `unlock()` was automatically called as part of `AllocLock.drop()`. I removed it though because it felt overkill. As such, I suppose this can go too.

Reply:
Less code is certainly good :)

Reply:
You could also use the `with` pattern, like [`LocalKey`](https://doc.rust-lang.org/std/thread/struct.LocalKey.html#method.with). This'd make the critical section, in effect, scoped and call sites couldn't forget to release the lock.

Could be considered over-engineering though...

Reply:
Probably at this stage, but maybe worth revisiting in the future